src/codegen/__future__/definePlugin.ts

```diff
@@ -0,0 +1,211 @@
+   */
+  modifyDefaults?(
+    defaults: ts.VariableDeclaration,
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Modify the `servers` value definition in the head of the new TS document
+   * Can be async
+   * @param servers - the AST of the `servers` definition
+   * @param fullTree - the full AST tree of the document
+   * @param document - the full OAS document
+   */
+  modifyServers?(
+    servers: ts.VariableDeclaration,
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Run before an OAS endpoint is converted to an AST node
+   * Can be async
+   * @param schema - the current endpoint Object from the OAS document
+   * @param fullTree - the full AST tree
+   * @param document - the full OAS document
+   */
+  beforeConvertEndpoint?(
+    schema: OpenAPIV3.OperationObject,
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Run after an OAS endpoint is converted to an AST node
+   * Can be async
+   * @param schema - the current endpoint Object from the OAS document
+   * @param functionTree - the AST node for the function definition
+   * @param fullTree - the full AST tree
+   * @param document - the full OAS document
+   */
+  afterConvertEndpoint?(
+    schema: OpenAPIV3.OperationObject,
+    functionTree: ts.FunctionDeclaration,
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Run before an OAS schema is converted to an AST node
+   * Can be async
+   * @param schema - the current OAS Schema about to be converted
+   * @param fullTree - the full AST tree
+   * @param document - the full OAS document
+   */
+  beforeConvertSchema?(
+    schema: SchemaObject | OpenAPIV3.ReferenceObject,
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Run after an OAS schema is converted to an AST node
+   * Can be async
+   * @param schema - the current OAS Schema that was converted
+   * @param typeTree - the AST Type Definition
+   * @param fullTree - the full AST tree
+   * @param document - the full OAS document
+   */
+  afterConvertSchema?(
+    schema: SchemaObject | OpenAPIV3.ReferenceObject,
+    typeTree: ts.TypeNode,
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Runs after the full AST tree has been created, but before the ts file contents have been generated
+   * Can be async
+   * @param fullTree - the full AST tree
+   * @param document - the full OAS document
+   */
+  afterASTGenerate?(
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Runs after the TS file contents have been generated, but before writing to file.
+   * Can be async
+   * @param fileName - name of the file to write to (should this be a full path object?)
```

Comment:
When we have the full path, yes. One could get the fileName when needed from this.

Reply:
i've updated the type of this.